sklearn/utils/extmath.py

```diff
@@ -195,6 +195,42 @@ def safe_sparse_dot(a, b, dense_output=False):
         return fast_dot(a, b)
 
 
+def _block_slices(dim_size, block_size):
+    """Generator that yields slice objects for indexing into sequential blocks
+    of an array along a particular axis
+    """
+    count = 0
+    while True:
+        yield slice(count, count + block_size, 1)
+        count += block_size
+        if count > dim_size:
+            raise StopIteration
+
+
+def _blockwise_dot(A, B, max_rows, max_cols, out=None):
+    """Computes the dot product of two matrices in a block-wise fashion. Only
+    blocks of `A` with a maximum size of `(max_rows, max_cols)` will be
+    processed simultaneously.
+    """
+    m,  n = A.shape
+    n1, o = B.shape
+    if n1 != n:
+        raise ValueError('matrices are not aligned')
```

Comment:
Please provide informative error messages with the actual values, for instance:

raise ValueError('Incompatible input array shapes to compute dot product: %r and %r'
                  % (A.shape, B.shape))